File: solutions/scala/complex-numbers/1/src/main/scala/ComplexNumber.scala
case class ComplexNumber(real: Double = 0, imaginary: Double = 0):
    def +(other: ComplexNumber): ComplexNumber =
        ComplexNumber(
            real + other.real,
            imaginary + other.imaginary
        )

    def -(other: ComplexNumber): ComplexNumber =
        ComplexNumber(
            real - other.real,
            imaginary - other.imaginary
        )

    def *(other: ComplexNumber): ComplexNumber =
        ComplexNumber(
            real * other.real - imaginary * other.imaginary,
            imaginary * other.real + real * other.imaginary
        )

    def /(other: ComplexNumber): ComplexNumber =
        val denominator =
            other.real * other.real + other.imaginary * other.imaginary

        ComplexNumber(
            (real * other.real + imaginary * other.imaginary) / denominator,
            (imaginary * other.real - real * other.imaginary) / denominator
        )

    def conjugate: ComplexNumber =
        ComplexNumber(real, -imaginary)

    def abs: Double =
        Math.sqrt(real * real + imaginary * imaginary)

object ComplexNumber:
    def exp(c: ComplexNumber): ComplexNumber =
        val expReal = Math.exp(c.real)
        ComplexNumber(
            expReal * Math.cos(c.imaginary),
            expReal * Math.sin(c.imaginary)
        )

File: solutions/java/complex-numbers/1/src/main/java/ComplexNumber.java
record ComplexNumber(double real, double imaginary) {

    double abs() {
        return Math.hypot(this.real, this.imaginary);
    }

    ComplexNumber add(final ComplexNumber other) {
        return new ComplexNumber(
                this.real + other.real,
                this.imaginary + other.imaginary
        );
    }

    ComplexNumber subtract(final ComplexNumber other) {
        return new ComplexNumber(
                this.real - other.real,
                this.imaginary - other.imaginary
        );
    }

    ComplexNumber multiply(final ComplexNumber other) {
        return new ComplexNumber(
                this.real * other.real - this.imaginary * other.imaginary,
                this.imaginary * other.real + this.real * other.imaginary
        );
    }

    ComplexNumber divide(final ComplexNumber other) {
        final double denominator = other.real * other.real
                                   + other.imaginary * other.imaginary;

        final double realNumerator = this.real * other.real
                                     + this.imaginary * other.imaginary;

        final double imaginaryNumerator = this.imaginary * other.real
                                          - this.real * other.imaginary;

        return new ComplexNumber(
                realNumerator / denominator,
                imaginaryNumerator / denominator
        );
    }

    ComplexNumber conjugate() {
        return new ComplexNumber(this.real, -this.imaginary);
    }

    ComplexNumber exponentialOf() {
        final double exponentialReal = Math.exp(this.real);

        return new ComplexNumber(
                exponentialReal * Math.cos(this.imaginary),
                exponentialReal * Math.sin(this.imaginary)
        );
    }
}